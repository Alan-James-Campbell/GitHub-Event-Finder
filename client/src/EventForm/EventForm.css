.eventFormInputs {
  margin-bottom: 15px;
}

.eventFormErrors {
  margin-top: -15px;
}

.submitAlert {
  font-size: 12px;
  color: red;
  margin-top: 10px;
}

.eventTypeCheckBoxes {
  margin: 25px;
  font-size: 12px;
  display: inline-block;
}


