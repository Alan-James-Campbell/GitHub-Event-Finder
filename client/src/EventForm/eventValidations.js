export const notEmpty = (value, a, b, name) => value && value.length > 0 ? undefined : `${name} cannot be empty`
