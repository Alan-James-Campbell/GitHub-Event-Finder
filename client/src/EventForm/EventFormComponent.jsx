import React 		    from 'react'
import { Field }    from 'redux-form'
import ReactTable   from 'react-table'
import { Button }   from 'react-bootstrap'
import { notEmpty } from './eventValidations'
import              'react-table/react-table.css'
import       		    './EventForm.css'

const EventForm = ({eventList, eventTypes, disableSubmit, submitError, getEvents, updateSubmitError, userName, repoName, valid, initialValues, isLoading}) => {

  return (
    <div className='eventForm'>
      <form onSubmit={e => getEvents(e, userName, repoName)}>
        
        <label>User Name</label>
        <Field
          name="userName" 
          label='User Name'
          component={renderField}
          onFocus={e => submitError ? updateSubmitError('') : null }
          validate={[notEmpty]}
        />
      
        <label>Repo Name</label>
        <Field 
          name="repoName" 
          label='Repo Name'
          component={renderField}
          onFocus={e => submitError ? updateSubmitError('') : null }
          validate={[notEmpty]}
        />

        {(eventTypes.length > 0)&&(
          eventTypes.map((type, idx) => {
          	return (
          	  <div className='eventTypeCheckBoxes' key={idx}>	
              <div className='checkBoxes'/>
  	            <label>{type}</label>	
  	            <Field name={type} component={renderField} type='checkbox' defaultChecked={true}/>
              </div>
          	)
          })
        )}
        
        <div>
          <Button
            disabled={disableSubmit  || !valid}
            type="submit"
          >
            {isLoading ? (
              <div className="spinner-border" role="status">
                <span className="sr-only"></span>
              </div>
             ) : (
              <div>
                Submit
              </div>
            )}
            
          </Button>
        </div>     

      </form>

      {submitError&&(
        <div className='submitAlert'>
          <div className="alert alert-danger" role="alert">
            {submitError}
          </div>
        </div>
      )}

      {(eventList.length > 0)&&(
        <div>
          <ReactTable
            style={{'width':'100%', 'margin':'20px 0', 'backgroundColor':'white', 'color':'black'}}
            data={eventList}
            filterable
            columns={[
              {
                columns: [
                  {
                    Header: 'Event Type',
                    accessor: 'type',
                    style: {'fontSize': '12px'},
                    sortMethod: (a, b) => {if (a === b) {return 0; } return a > b ? 1 : -1; }, 
                    filterMethod: (filter, row) => row[filter.id].toLowerCase().startsWith(filter.value.toLowerCase())
                  },                 
                  {
                    Header: 'User Name',
                    accessor: 'actor.display_login',
                    style: {'fontSize': '12px'},
                    sortable: false,
                    filterable: false
                    //user information is uniform and does not need sort or filter options           
                  },                      
                  {
                    Header: 'User Id',
                    accessor: 'actor.id',
                    style: {'fontSize': '12px'},
                    sortable: false,
                    filterable: false,
                    //user information is uniform and does not need sort or filter options           
                  },
                  {
                    Header: 'TimeStamp',
                    accessor: 'created_at',
                    style: {'fontSize': '12px'},
                    sortMethod: (a, b) => {
                      const a1 = new Date(a)
                      const b1 = new Date(b)
                      if (a1 === b1) {return 0; } return a1 > b1 ? 1 : -1;
                    }
                  },              

                ]
              }
            ]}
            defaultPageSize={10}
            defaultSorted={[{ id: 'event', desc: true } ]} 
            className="-striped -highlight"
          />
          
          <div style={{ textAlign: "center" }}>
            <em>Tip: Type in boxes to filter content!</em><br/>
            <em>Tip: Hold shift when sorting to multi-sort!</em>
          </div>

        </div>
      )}

    </div>
  ) 
}

const renderField = ({ input, label, type, meta: { touched, error, warning } }) => (
  <div >
    <input className='eventFormInputs'{...input} placeholder={label} type={type}/>  
    {touched && (error && <div className='eventFormErrors'><small style={{'color':'red'}}>{error}</small></div>)}
  </div>
)

export default EventForm
