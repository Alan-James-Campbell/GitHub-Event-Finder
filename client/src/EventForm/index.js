import EventFormComponent   from './EventFormComponent'
import {connect}     		from 'react-redux'
import { reduxForm }        from 'redux-form'
import _                    from 'lodash'
import {fetchEvents} 		from '../reducers/events'

const mapStateToProps = state => {
 
  const onlyUnique = (value, index, self) => self.indexOf(value) === index
  
  const createEventTypeObject = eventTypes => {
  	const eventTypeObject = {}
  	eventTypes.forEach(event => {
  	  const bool = _.get(state, `form.EventForm.values.${event}`, false)
      eventTypeObject[event] = bool
  	})
  	return eventTypeObject
  }

  const userName = _.get(state, 'form.EventForm.values.userName', '')
  const repoName = _.get(state, 'form.EventForm.values.repoName', '')
  let eventList   = _.get(state, 'events.eventList', [])
  let eventTypes = eventList.length > 0 ? eventList.map(event => event.type).filter(onlyUnique) : []
  const initialValues = Object.assign(
    {},
    (userName ? {userName} : {}),
    (repoName ? {repoName} : {}),
    (createEventTypeObject(eventTypes))
  )
  const {lastSubmittedUserName, lastSubmittedRepoName, hasSubmitted} = state.events
  const hasChangedAfterSubmission = (hasSubmitted) && ((lastSubmittedUserName !== userName || lastSubmittedRepoName !== repoName))
  eventList = eventList.filter(event => (Object.keys(initialValues).includes(event.type)) && (initialValues[event.type] === true))
  const disableSubmit = (hasSubmitted && !hasChangedAfterSubmission)
  eventTypes = disableSubmit ? eventTypes : []
  eventList = disableSubmit ? eventList : []
  
  return {
    userName,
    repoName,
    eventList,
    eventTypes,
    initialValues,
    disableSubmit
  }
}

const mapDispatchToProps = (dispatch,ownProps) => {
  return {
    getEvents(e, userName, repoName) {
      e.preventDefault()
      dispatch(fetchEvents(userName, repoName))
    }
  }
}

const EventForm = reduxForm({
  form: 'EventForm',
  enableReinitialize: true,
})(EventFormComponent)

export default connect(mapStateToProps, mapDispatchToProps)(EventForm)
