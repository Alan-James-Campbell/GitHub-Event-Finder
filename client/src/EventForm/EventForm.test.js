import React       from 'react'
import { shallow } from 'enzyme'
import toJson      from 'enzyme-to-json'
import EventForm from './EventFormComponent'
import { Field }    from 'redux-form'

    
describe('EventForm', () => {
	const props = {
      eventList: [],
	  eventTypes: [],
	  disableSubmit: false, 
	  submitError: '', 
	  getEvents: function(){},
	  updateSubmitError: function(){},
	  userName: '',
	  repoName: '',
	  valid: true,
	  initialValues: {},
	  isLoading: false
    }

    it('renders without crashing given the required props', () => {
      const wrapper = shallow(<EventForm {...props} />)
      expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('does not render Event Type checkboxes when eventType list is empty', () => {
      const wrapper = shallow(<EventForm {...props} />)
      expect(wrapper.contains(<div className='unique' />)).toEqual(false)
    })    

    it('renders Event Type checkboxes when eventType list is populated', () => {
      const newProps = Object.assign({}, props, {eventTypes: ['PushEvent', 'CreateEvent']})
      const wrapper = shallow(<EventForm {...newProps} />)
      expect(wrapper.contains(<div className='checkBoxes' />)).toEqual(true)
    })
})