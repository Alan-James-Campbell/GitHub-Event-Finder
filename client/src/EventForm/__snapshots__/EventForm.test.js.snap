// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm renders without crashing given the required props 1`] = `
<div
  className="eventForm"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      User Name
    </label>
    <Field
      component={[Function]}
      label="User Name"
      name="userName"
      onFocus={[Function]}
      validate={
        Array [
          [Function],
        ]
      }
    />
    <label>
      Repo Name
    </label>
    <Field
      component={[Function]}
      label="Repo Name"
      name="repoName"
      onFocus={[Function]}
      validate={
        Array [
          [Function],
        ]
      }
    />
    <div>
      <Button
        active={false}
        disabled={false}
        type="submit"
        variant="primary"
      >
        <div>
          Submit
        </div>
      </Button>
    </div>
  </form>
</div>
`;
